src/Orleans.Transactions/State/TransactionQueue.cs

```diff
@@ -347,6 +354,24 @@ public void NotifyOfCancel(Guid transactionId, DateTime timeStamp, Transactional
         /// called on activation, and when recovering from storage conflicts or other exceptions.
         /// </summary>
         public async Task NotifyOfRestore()
+        {
+            await Ready();
+            this.restoreTask = Restore();
+            await this.restoreTask;
+        }
+
+        /// <summary>
+        /// called on activation, and when recovering from storage conflicts or other exceptions.
+        /// </summary>
+        public async Task Ready()
+        {
+            if (!this.restoreTask.IsCompleted)
```

Comment:
If `Restore` fails (eg `storage.Load()` exception) then I imagine we'd need to restart recovery on the next call.

This logic will instead ignore failures on subsequent calls.
That seems incorrect/dangerous to me.

For example, this could manifest after a write conflict.

Reply:
First, this solution sucks, but addresses the problem until we can refactor the call patterns.

The issue encountered was that in recovery scenarios, we clear the locks and commit logs, then reload and recover.  The reload is an async operation, surrendering the grains turn.  This allows new reentrant calls to be processed while the state is empty and still being reloaded. 

By holding the restore task and awaiting it before any incoming reentrant call we block all such calls while the queue is restoring.  This call does not need to trigger anything, just delay new operations until any pending restore is complete.

In the case where there is no restore error, these awaits should be fast as they will mostly occur on completed tasks.  In the case where there IS a restore error, all the pending calls will fail, which is good.  The internal recovery logic is responsible for retrying the restore, not the Ready calls, they just make sure it’s complete before allowing new requests.

Reply:
I have an idea for a simpler solution - we can use the reader-writer lock to make callers wait for the load to complete. I will try that out.